File: src/main/java/com/nugurang/service/ProjectService.java
package com.nugurang.service;

import com.nugurang.constant.RoleName;
import com.nugurang.dao.EventDao;
import com.nugurang.dao.ProjectDao;
import com.nugurang.dao.RoleDao;
import com.nugurang.dao.TeamDao;
import com.nugurang.dao.XrefUserProjectDao;
import com.nugurang.dto.ProjectInputDto;
import com.nugurang.entity.ProjectEntity;
import com.nugurang.entity.XrefUserProjectEntity;
import java.util.Optional;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class ProjectService {
    private final EventDao eventDao;
    private final RoleDao roleDao;
    private final ProjectDao projectDao;
    private final TeamDao teamDao;
    private final UserService userService;
    private final XrefUserProjectDao xrefUserProjectDao;

    @Transactional
    public ProjectEntity createProject(ProjectInputDto projectInputDto, Long teamId) {
        ProjectEntity projectEntity = projectDao.save(ProjectEntity.builder().name(projectInputDto.getName()).finished(false).team(teamDao.findById(teamId).get()).event(Optional.ofNullable(projectInputDto.getEvent()).flatMap(id -> eventDao.findById(id)).orElse(null)).build());
        xrefUserProjectDao.save(XrefUserProjectEntity.builder().user(userService.getCurrentUser().get()).project(projectEntity).role(roleDao.findByName(RoleName.OWNER.name())).build());
        return projectEntity;
    }

    public Optional<ProjectEntity> getProject(Long projectId) {
        return projectDao.findById(projectId);
    }

    @Transactional
    public ProjectEntity updateProject(ProjectInputDto projectInputDto, Long projectId) {
        return projectDao.save(projectDao.findById(projectId).map(projectEntity -> {
            projectEntity.setName(projectInputDto.getName());
            projectEntity.setEvent(Optional.ofNullable(projectInputDto.getEvent()).flatMap(eventId -> eventDao.findById(eventId)).orElse(null));
            return projectEntity;
        }).get());
    }

    public void deleteProject(Long projectId) {
        projectDao.deleteById(projectId);
    }

    //<editor-fold defaultstate="collapsed" desc="delombok">
    @SuppressWarnings("all")
    
    public ProjectService(final EventDao eventDao, final RoleDao roleDao, final ProjectDao projectDao, final TeamDao teamDao, final UserService userService, final XrefUserProjectDao xrefUserProjectDao) {
        this.eventDao = eventDao;
        this.roleDao = roleDao;
        this.projectDao = projectDao;
        this.teamDao = teamDao;
        this.userService = userService;
        this.xrefUserProjectDao = xrefUserProjectDao;
    }
    //</editor-fold>
}

File: src/main/kotlin/com/nugurang/service/ProjectService.kt
package com.nugurang.service

import com.nugurang.annotation.DaoOp
import com.nugurang.constant.RoleName
import com.nugurang.dao.*
import com.nugurang.dto.ProjectInputDto
import com.nugurang.entity.ProjectEntity
import com.nugurang.entity.RoleEntity
import com.nugurang.entity.TeamEntity
import com.nugurang.entity.XrefUserProjectEntity
import com.nugurang.exception.NotFoundException
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.validation.annotation.Validated
import javax.validation.Valid

@Validated
@Service
class ProjectService(
    private val eventDao: EventDao,
    private val roleDao: RoleDao,
    private val projectDao: ProjectDao,
    private val teamDao: TeamDao,
    private val userService: UserService,
    private val xrefUserProjectDao: XrefUserProjectDao
) {

    @DaoOp
    fun createProject(@Valid projectInputDto: ProjectInputDto, teamId: Long): ProjectEntity {
        val projectEntity = projectDao.save(
            ProjectEntity(
                name = projectInputDto.name,
                finished = false,
                team = teamDao.findByIdOrNull(teamId) ?: throw NotFoundException(TeamEntity::class.java),
                event = projectInputDto.event?.let(eventDao::findByIdOrNull)
            )
        )

        xrefUserProjectDao.save(
            XrefUserProjectEntity(
                user = userService.getCurrentUser(),
                project = projectEntity,
                role = roleDao.findByName(RoleName.OWNER.name)
                    ?: throw NotFoundException(RoleEntity::class.java)
            )
        )
        return projectEntity
    }

    fun getProject(projectId: Long): ProjectEntity {
        return projectDao.findByIdOrNull(projectId) ?: throw NotFoundException(ProjectEntity::class.java)
    }

    @DaoOp
    fun updateProject(@Valid projectInputDto: ProjectInputDto, projectId: Long): ProjectEntity {
        return projectDao.save(
            projectDao.findByIdOrNull(projectId)?.let { projectEntity ->
                projectEntity.name = projectInputDto.name
                projectEntity.event = projectInputDto.event?.let(eventDao::findByIdOrNull)
                projectEntity
            } ?: throw NotFoundException(ProjectEntity::class.java)
        )
    }

    @DaoOp
    fun deleteProject(projectId: Long) {
        projectDao.deleteById(projectId)
    }
}